lib/bap_llvm/llvm_binary.hpp

```diff
@@ -185,150 +182,161 @@ struct section {
+    for (section_iterator it : *obj_secs) {
+        const coff_section *s = getCOFFSection(obj, it);
+        secs->push_back(make_section(*s, *image_base));
     }
-    return sections;
+    return secs;
 }
 
 } //namespace sec
 
 namespace img {
+
+using std::move;
 using namespace llvm;
 using namespace llvm::object;
 
 struct image {
-    virtual uint64_t entry() const = 0;
-    virtual const std::string &arch() const = 0;
-    virtual const std::vector<seg::segment>& segments() const = 0;
-    virtual const std::vector<sym::symbol>& symbols() const = 0;
-    virtual const std::vector<sec::section>& sections() const = 0;
-    virtual ~image() {}
+    std::string arch;
+    uint64_t entry;
+    seg::segments segments;
+    sym::symbols  symbols;
+    sec::sections sections;
 };
 
 std::string image_arch(const ObjectFile& obj) {
     return Triple::getArchTypeName(static_cast<Triple::ArchType>(obj.getArch()));
 }
 
 template <typename ELFT>
-uint64_t image_entry(const ELFObjectFile<ELFT>& obj) {
-    return obj.getELFFile()->getHeader()->e_entry;
+error_or<uint64_t> image_entry(const ELFObjectFile<ELFT>& obj) {
+    return error_or<uint64_t>(obj.getELFFile()->getHeader()->e_entry);
 }
 
-uint64_t image_entry(const MachOObjectFile &obj) {
+error_or<uint64_t> image_entry(const MachOObjectFile &obj) {
     return image_entry_macho(obj);
 }
 
-uint64_t image_entry(const COFFObjectFile &obj) {
+error_or<uint64_t> image_entry(const COFFObjectFile &obj) {
     return image_entry_coff(obj);
 }
 
 template <typename T>
-struct objectfile_image : image {
-    explicit objectfile_image(std::unique_ptr<T> ptr)
-	: arch_(image_arch(*ptr))
-	, entry_(image_entry(*ptr))
-	, segments_(seg::read(*ptr))
-	, symbols_(sym::read(*ptr))
-	, sections_(sec::read(*ptr))
-	, binary_(move(ptr))
-    {}
-    const std::string &arch() const { return arch_; }
-    uint64_t entry() const { return entry_; }
-    const std::vector<seg::segment>& segments() const { return segments_; }
-    const std::vector<sym::symbol>& symbols() const { return symbols_; }
-    const std::vector<sec::section>& sections() const { return sections_; }
-protected:
-    const std::string arch_;
-    uint64_t entry_;
-    std::vector<seg::segment> segments_;
-    std::vector<sym::symbol> symbols_;
-    std::vector<sec::section> sections_;
-    std::unique_ptr<T> binary_;
-};
+image* create_image(error_or<object::Binary> &binary) {
+    if (auto ptr = llvm::dyn_cast<T>(binary.get())) {
+        auto arch     = image_arch(*ptr);
+        auto entry    = image_entry(*ptr);
+        auto segments = seg::read(*ptr);
+        auto symbols  = sym::read(*ptr);
+        auto sections = sec::read(*ptr);
+
+        if (auto er = entry || segments || symbols || sections) {
+            bap_notify_error(er.message());
+            return nullptr;
+        }
+        binary << entry.warnings() << segments.warnings() << symbols.warnings() << sections.warnings();
+        for (auto w : binary.warnings())
```

Comment:
you should print warnings only if BAP_DEBUG is set to 1. This is temporary, of course, as later with the new interface a frontend will take care of warnings.